src/shared_realm.cpp

```diff
@@ -484,11 +496,19 @@ static void check_write(const Realm* realm)
     if (realm->config().immutable() || realm->config().read_only_alternative()) {
         throw InvalidTransactionException("Can't perform transactions on read-only Realms.");
     }
+    if (realm->is_frozen()) {
+        throw InvalidTransactionException("Can't perform transactions on a frozen Realm");
+    }
+    if (!realm->is_closed() && realm->get_number_of_versions() > realm->config().max_number_of_active_versions) {
+        throw InvalidTransactionException(util::format("Number of active versions (%1) in the Realm exceeded the limit of %2",
+                realm->get_number_of_versions(),
+                realm->config().max_number_of_active_versions));
+    }
 }
 
 void Realm::verify_thread() const
 {
-    if (!m_execution_context.contains<std::thread::id>())
+    if (!m_execution_context.contains<std::thread::id>() || m_frozen_version)
```

Comment:
Checking m_frozen_version first is a small but free perf gain as getting the current thread id is a little bit of work.